### PR TITLE
Correctly stack debuginfo for inlined body in classic mode

### LINKED ISSUE #1121: Rewrite Debuginfo values in classic mode so inlined frames are shown correctly

### PR DESCRIPTION
**Pull request by Gbury**
Should close #1121 .

We introduce a new nullary primitive, that is isnerted at the beginning of every function body inlined when in classic mode. This primitive is then used in to_cmm to correctly concatenate/stack the debuginfos inside of the inlined bodies. To do so, `to_cmm_env` is extended to record the debuginfo of the currently inlined body (if any), and continuations stored in the env for inlining are stored together with the current inlined debuginfo, so that it can be restored when the continuation is inlined. `to_cmm` can then add the inlined debuginfo to all the debuginfos that are used in the generated cmm.

### CODE REVIEW

**Comment by mshinwell** [comment]
I will review this

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (mshinwell) [inline middle_end/flambda2/to_cmm/to_cmm_expr.ml:562]: If the debuginfo was "none", we probably shouldn't add the inlined debuginfo - might be confusing?

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (mshinwell) [inline middle_end/flambda2/to_cmm/to_cmm_expr.ml:611]: Same as above

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (mshinwell) [inline middle_end/flambda2/to_cmm/to_cmm_expr.ml:679]: Same as above (if we do change this, maybe leave CRs noting that we need to call add_inlined_debuginfo when we do end up with the correct dbg one day)

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (mshinwell) [inline middle_end/flambda2/to_cmm/to_cmm_expr.ml:758]: Maybe handler_body_dbg might be clearer?  (bit unclear at the moment whether this is just part of the debuginfo or all of it, and which expression it applies to...)

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (mshinwell) [inline middle_end/flambda2/to_cmm/to_cmm_expr.ml:285]: Here and below there are various cases all of which come from Apply_cont expressions (presumably via the apply_cont function below?).  If so, it would be clearer to put this add_inlined_debuginfo call right at the top of the apply_cont function, instead of in these various other places.

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: middle_end/flambda2/to_cmm/to_cmm_env.mli
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_env.mli:102]: fixed

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_expr.ml:37]: I changed the parameter name to `dbg_with_inlined`, and adjusted the callsites

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_expr.ml:562]: Indeed, I reverted back to using `Debuginfo.non` with a CR to remember to apply `Env.add_inlined_debuginfo` when a proper debuginfo is available.

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_expr.ml:611]: fixed

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_expr.ml:679]: fixed

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_expr.ml:758]: renamed to `handler_body_inlined_debuginfo`

File: middle_end/flambda2/to_cmm/to_cmm_env.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_env.ml:168]: fixed

File: middle_end/flambda2/to_cmm/to_cmm_env.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_env.ml:256]: fixed

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_expr.ml:235]: done

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_expr.ml]: done, these function now take a `~dbg_with_inlined` parameter

**Comment by Gbury** [comment]
Note: before merging this PR, we should wait for #1161 to be merged, and then i'll rebase this PR, and add tests to the  `backtracte` folder (notably change some tests to also test classic mode, and potentially add some tests for cross-module inlining between classic mode and simplify).

**Review by mshinwell — APPROVED** [review]
Code looks good now, just need tests as discussed

**Review by mshinwell — CHANGES_REQUESTED** [review]
Just a couple of minor things

**Review by mshinwell — CHANGES_REQUESTED** [review]
Just a few minor things

**Review by mshinwell — APPROVED** [review]

**Comment by Gbury** [comment]
This has been rebased on `main`, and new tests have been added to the testsuite, see the `backtrace-mutlifiles` folder. Waiting for CI because it's likely that the new test is not working on e.g. closure.

File: ocaml/testsuite/tests/backtrace-multifiles/main.ml
Comment (mshinwell) [inline ocaml/testsuite/tests/backtrace-multifiles/main.ml:5]: Presumably these CRs are to be addressed before merging?

Please also add a comment explaining what the test does, as it's kind of complicated to look at.

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: ocaml/testsuite/tests/backtrace-multifiles/main.ml
Comment (Gbury) [reply ocaml/testsuite/tests/backtrace-multifiles/main.ml:5]: Actually, these CR are because of #1199 , so I think it's better to merge this PR with them.

I'll add some documentation to the test to explain what it tests.

Comment (Gbury) [reply ocaml/testsuite/tests/backtrace-multifiles/main.ml]: I've added a comment at the end of `main.ml` to explain the goal of the test.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**
